[plugins] EnvGen: fix initialization

mtmccrea opened the pull request:
<!-- Please see CONTRIBUTING.md for guidelines. -->

## Purpose and Motivation
This fixes the initialization of `EnvGen` as well as improves (but doesn't fully fix) a segment duration bug, #6180.

Additionally:
- unit tests added to check initialization by the first output sample.
- fix an `Env.circle` startup behavior

This PR is part of an ongoing effort to [fix UGen initializations](https://github.com/supercollider/rfcs/pull/19), the progress of which is [tracked here](https://github.com/mtmccrea/supercollider/wiki/UGen-Fix-List).

**Note:** I did some cleanup of extraneous print statements and stray comments, as a standalone commit. If this isn't preferred, I can remove it. I think it's helpful to clean up as we go.

<!-- If this fixes an open issue, link to it by writing "Fixes #555." -->


## Types of changes

<!-- Delete lines that don't apply -->

- Bug fixes
- Breaking change - changes are technically breaking for the first output sample (or, e.g., a one-sample phase change from previously)

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested
- [x] All tests are passing
  - ~~the state of the CoreUGens unit test is the same as reported in https://github.com/supercollider/supercollider/pull/6035 (all relevant tests passing)~~
  - ~~`TestEnvGen_server:test_shapeHold_setEndValue` fails but is fixed by fixed by #6185~~
- [ ] Documentation updated: n/a
- [x] This PR is ready for review

mossheim commented:
@mtmccrea what's the intention with these PRs since the corresponding RFC hasn't been merged yet?

mtmccrea commented:
The RFC was used as a space for corralling ideas from a number of issues and side threads that had gone on for a long time before leading up to that. So each PR is intended to be reviewed like any other, with the RFC serving as a contextual backdrop so I don't have to explain it with every PR.

The intention of the RFC was also to motivate people to fix this pervasive problem, less a question of _whether_ UGen initialization needed to be fixed (I think there's wide agreement on that). But so far I've been the only one working on it...

mtmccrea commented:
Note for the Reviewer:
In `EnvGen`, I found that resetting the state of `unit->m_counter` back to `0` could cause the synth not to release (for `doneAction: 2`), so when I'm resetting the UGen state at the tail of the Ctor, I don't reset `m_counter`. I suspect it may have something to do with the fact that `next_k` (which is called for the init sample calculation) decrements the counter:
https://github.com/supercollider/supercollider/blob/8338b0c55cc5d1be73341b09266c45abbb2ab8cd/server/plugins/LFUGens.cpp#L2822

I'd prefer to have a better grasp on why that is the case, and I'll continue looking into it. In any case it would be good if the Reviewer would also take a look at this aspect.

mossheim commented:
> @mossheim The RFC was used as a space for corralling ideas from a number of issues and side threads that had gone on for a long time before leading up to that. So each PR is intended to be reviewed like any other, with the RFC serving as a contextual backdrop so I don't have to explain it with every PR.
> 
> The intention of the RFC was also to motivate people to fix this pervasive problem, less a question of _whether_ UGen initialization needed to be fixed (I think there's wide agreement on that). But so far I've been the only one working on it.

Got it thanks! I will take a look at them sometime in the next weeks then if someone hasn't already.

mtmccrea commented:
In the course of fixing another EnvGen unit test, I decided to roll that, as well as the doc changes, into separate PRs. They don't need to wait on this PR to resolve.

mtmccrea commented:
Failing test will pass once #6185 is merged (why didn't this fail on Windows? 🤔)

mtmccrea commented:
Rabased now tests are all passing

telephon commented:
Should we have some more tests for envelope?

dyfer commented:
@mtmccrea thanks for this! Are you considering adding tests for this or would should we pull this in as is?

mtmccrea commented:
Thanks for following up with this @dyfer. I'm not actively contributing at the moment, so I don't think I'd be able to add tests soon. So it would be great to see this pulled in, and I can address tests as a separate issue in the future, assuming the current work is acceptable.

mtmccrea commented:
And FWIW, there were [2 tests updated](https://github.com/supercollider/supercollider/pull/6185) in the course of making this fix, which, while not testing specifically this change, hopefully offers some assurance that EnvGen is implicitly tested throughout many unit tests.

mtmccrea commented:
I've now added a few tests for the initialization of `EnvGen`. 
I also tested the `circle` methods, which I found had broken when the interface to `Delay1.kr` had changed, so that's fixed and docs touched up as well.

wrt the env duration bug, #6180, this PR _mostly_ fixes it, though there's an edge case regarding the last sample of the last segment, which should be handled in a separate PR, since this one is concerned with proper initialization.

mtmccrea commented:
There was a spurious test timeout failure in the pitch tracker (macOS), which I can't recreate on my machine and doesn't fail after re-running, so I think it's ok...

Otherwise I think I've addressed all the issues raised by commenters so far and it's ready for review.

telephon reviewed (approved):
Thanks, this is a very important fix.
